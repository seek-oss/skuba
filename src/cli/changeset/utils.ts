// Adapted from https://github.com/changesets/action/blob/21240c3cd1d2efa2672d64e0235a03cf139b83e6/src/gitUtils.ts

import type { Package } from '@manypkg/get-packages';
import { getPackages } from '@manypkg/get-packages';
import execa from 'execa';
import mdastToString from 'mdast-util-to-string';
import remarkParse from 'remark-parse';
import remarkStringify from 'remark-stringify';
import { unified } from 'unified';

export const BumpLevels = {
  dep: 0,
  patch: 1,
  minor: 2,
  major: 3,
} as const;

export async function getVersionsByDirectory(cwd: string) {
  const { packages } = await getPackages(cwd);
  return new Map(packages.map((x) => [x.dir, x.packageJson.version]));
}

export async function getChangedPackages(
  cwd: string,
  previousVersions: Map<string, string>,
) {
  const { packages } = await getPackages(cwd);
  const changedPackages = new Set<Package>();

  for (const pkg of packages) {
    const previousVersion = previousVersions.get(pkg.dir);
    if (previousVersion !== pkg.packageJson.version) {
      changedPackages.add(pkg);
    }
  }

  return [...changedPackages];
}

export function getChangelogEntry(changelog: string, version: string) {
  const ast = unified().use(remarkParse).parse(changelog);

  let highestLevel: number = BumpLevels.dep;

  const nodes = ast.children;
  let headingStartInfo:
    | {
        index: number;
        depth: number;
      }
    | undefined;
  let endIndex: number | undefined;

  for (let i = 0; i < nodes.length; i++) {
    const node = nodes[i];
    if (node.type === 'heading') {
      const stringified: string = mdastToString.toString(node);
      const match = /(major|minor|patch)/.exec(stringified.toLowerCase());
      if (match !== null) {
        const level = BumpLevels[match[0] as 'major' | 'minor' | 'patch'];
        highestLevel = Math.max(level, highestLevel);
      }
      if (headingStartInfo === undefined && stringified === version) {
        headingStartInfo = {
          index: i,
          depth: node.depth,
        };
        continue;
      }
      if (
        endIndex === undefined &&
        headingStartInfo !== undefined &&
        headingStartInfo.depth === node.depth
      ) {
        endIndex = i;
        break;
      }
    }
  }
  if (headingStartInfo) {
    ast.children = ast.children.slice(headingStartInfo.index + 1, endIndex);
  }
  return {
    content: unified().use(remarkStringify).stringify(ast),
    highestLevel,
  };
}

export const execWithOutput = async (
  command: string,
  args?: string[],
  options: { ignoreReturnCode?: boolean; cwd?: string } = {},
) => {
  try {
    const { exitCode, stdout, stderr } = await execa(
      command,
      args,
      ...(options.cwd ? [{ cwd: options.cwd }] : []),
    );

    return {
      code: exitCode,
      stdout,
      stderr,
    };
  } catch ({ exitCode, stdout, stderr }) {
    if (!options.ignoreReturnCode) {
      throw new Error(stderr as string);
    }

    return {
      code: exitCode,
      stdout,
      stderr,
    };
  }
};

export function sortTheThings(
  a: { private: boolean; highestLevel: number },
  b: { private: boolean; highestLevel: number },
) {
  if (a.private === b.private) {
    return b.highestLevel - a.highestLevel;
  }
  if (a.private) {
    return 1;
  }
  return -1;
}
