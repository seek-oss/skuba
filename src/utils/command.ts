import path from 'path';

export type Command = (typeof COMMAND_LIST)[number];

export const COMMAND_ALIASES: Record<string, Command> = {
  '-h': 'help',
  '--help': 'help',
  '-v': 'version',
  '--version': 'version',
};

export const COMMAND_DIR = path.join(__dirname, '..', 'cli');

export const COMMAND_LIST = [
  'build',
  'build-package',
  'configure',
  'format',
  'help',
  'init',
  'lint',
  'node',
  'release',
  'start',
  'test',
  'version',
] as const;

export const COMMAND_SET = new Set<string>(COMMAND_LIST);

export const commandToModule = (command: Command): string =>
  command
    .split('-')
    .map((segment, index) =>
      index === 0
        ? segment
        : `${segment[0].toLocaleUpperCase()}${segment.slice(1)}`,
    )
    .join('');
