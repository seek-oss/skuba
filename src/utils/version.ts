import latestVersion from 'latest-version';

import { getSkubaManifest } from './manifest';
import { withTimeout } from './wait';

const latestSkubaVersion = async (): Promise<string | null> => {
  try {
    const result = await withTimeout(latestVersion('skuba'), { s: 2 });

    return result.ok ? result.value : null;
  } catch {
    return null;
  }
};

export const getSkubaVersion = async (): Promise<string> => {
  const { version } = await getSkubaManifest();

  return version;
};

type SkubaVersionInfo =
  | {
      isStale: true;

      local: string;
      latest: string;
    }
  | {
      isStale: false;

      local: string;
      latest: string | null;
    };

export const getSkubaVersionInfo = async (): Promise<SkubaVersionInfo> => {
  const [local, latest] = await Promise.all([
    getSkubaVersion(),
    latestSkubaVersion(),
  ]);

  if (latest === null) {
    // Assume we're up to date if we can't reach the npm registry
    return {
      isStale: false,
      local,
      latest,
    };
  }

  return {
    isStale: latest !== local,
    local,
    latest,
  };
};
