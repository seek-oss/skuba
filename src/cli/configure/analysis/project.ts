import path from 'path';

import fs from 'fs-extra';

import { buildPatternToFilepathMap, crawlDirectory } from '../../../utils/dir';
import { isErrorWithCode } from '../../../utils/error';
import { loadModules } from '../modules';
import { FileDiff, Files, Module, Options } from '../types';

import { determineOperation } from './diff';

export const createDestinationFileReader = (root: string) => async (
  filename: string,
): Promise<string | undefined> => {
  try {
    return await fs.readFile(path.join(root, filename), 'utf8');
  } catch (err: unknown) {
    if (isErrorWithCode(err, 'ENOENT')) {
      return;
    }

    throw err;
  }
};

const loadModuleFiles = async (modules: Module[], destinationRoot: string) => {
  const readDestinationFile = createDestinationFileReader(destinationRoot);

  const allFilepaths = await crawlDirectory(destinationRoot);

  const patterns = [...new Set(modules.flatMap((m) => Object.keys(m)))];

  const patternToFilepaths = buildPatternToFilepathMap(patterns, allFilepaths);

  const matchedFilepaths = [
    ...new Set(Object.values(patternToFilepaths).flat()),
  ];

  const fileEntries = await Promise.all(
    matchedFilepaths.map(
      async (filepath) =>
        [filepath, await readDestinationFile(filepath)] as const,
    ),
  );

  return {
    inputFiles: Object.fromEntries(fileEntries),
    patternToFilepaths,
  };
};

const processTextFiles = (
  modules: Module[],
  inputFiles: Readonly<Files>,
  patternToFilepaths: Record<string, string[]>,
) => {
  const outputFiles = { ...inputFiles };

  const textProcessorEntries = modules.flatMap((module) =>
    Object.entries(module).flatMap(([pattern, processText]) => {
      // Include the raw pattern along with any matched filepaths.
      // Some modules create a new file at the specified pattern.
      const filepaths = [pattern, ...(patternToFilepaths[pattern] ?? [])];

      return [...new Set(filepaths)].map(
        (filepath) => [filepath, processText] as const,
      );
    }),
  );

  for (const [filepath, processText] of textProcessorEntries) {
    outputFiles[filepath] = processText(
      outputFiles[filepath],
      outputFiles,
      inputFiles,
    );
  }

  return outputFiles;
};

export const diffFiles = async (opts: Options): Promise<FileDiff> => {
  const modules = await loadModules(opts);

  const { inputFiles, patternToFilepaths } = Object.freeze(
    await loadModuleFiles(modules, opts.destinationRoot),
  );

  const outputFiles = processTextFiles(modules, inputFiles, patternToFilepaths);

  const diffEntries = Object.entries(outputFiles)
    .filter(([filepath, data]) => inputFiles[filepath] !== data)
    .map(([filepath, data]) => {
      const operation = determineOperation(inputFiles[filepath], data);

      return [filepath, { data, operation }] as const;
    });

  return Object.fromEntries(diffEntries);
};
