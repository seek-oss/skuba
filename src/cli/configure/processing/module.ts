import { createStringReplacer } from '../../../utils/copy';

export const replacePackageReferences = (props: {
  old: {
    packageName: string;
    repoSlug: string;
  };
  new: {
    packageName: string;
    repoSlug: string;
  };
}) =>
  createStringReplacer([
    {
      input: new RegExp(`(['"])${props.old.packageName}(['"/])`, 'g'),
      output: `$1${props.new.packageName}$2`,
    },
    {
      input: new RegExp(props.old.repoSlug, 'ig'),
      output: props.new.repoSlug,
    },
  ]);
