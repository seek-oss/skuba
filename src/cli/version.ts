import { getSkubaVersion } from '../utils/version';

export const version = async () => {
  const skubaVersion = await getSkubaVersion();

  console.log(skubaVersion);
};
