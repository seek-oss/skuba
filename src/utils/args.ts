import assert from 'assert';

import { COMMAND_ALIASES } from './command';

/**
 * Parse process arguments.
 *
 * This function mutates the input list by removing the command name element.
 * Downstream commands can access their args with an ordinary
 * `process.argv.slice(2)`.
 *
 * Example input:
 *
 * ```typescript
 * ['/bin/node', 'node_modules/.bin/skuba', 'test', '--foo', '--bar']
 * ```
 */
export const parseArgs = (args = process.argv) => {
  const skubaIdx = args.findIndex((chunk) => /skuba(\.[jt]s)?$/.test(chunk));

  assert(skubaIdx >= 0, 'Cannot parse args for `skuba`');

  const rawCommand = (args[skubaIdx + 1] ?? 'help').toLocaleLowerCase();

  const commandName = COMMAND_ALIASES[rawCommand] ?? rawCommand;

  const payload = {
    commandName,
    args: args.slice(skubaIdx + 2),
  };

  args.splice(skubaIdx + 1, 1);

  return payload;
};
