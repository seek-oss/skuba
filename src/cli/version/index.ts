import { log } from '../../utils/logging.js';
import { getSkubaVersion } from '../../utils/version.js';

export const version = async () => {
  const skubaVersion = await getSkubaVersion();

  log.plain(skubaVersion);
};
