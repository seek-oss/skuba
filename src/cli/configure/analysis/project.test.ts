import { getSkubaVersion } from '../../../utils/version';
import { defaultOpts } from '../testing/module';

import * as project from './project';
import { diffFiles } from './project';

describe('diffFiles', () => {
  it('works from scratch', async () => {
    jest
      .spyOn(project, 'createDestinationFileReader')
      .mockReturnValue(() => Promise.resolve(undefined));

    const [outputFiles, version] = await Promise.all([
      diffFiles(defaultOpts),
      getSkubaVersion(),
    ]);

    const manifest = outputFiles['package.json'];

    manifest.data = manifest.data?.replace(
      new RegExp(version, 'g'),
      '0.0.0-semantically-released',
    );

    expect(outputFiles).toMatchSnapshot();
  });
});
