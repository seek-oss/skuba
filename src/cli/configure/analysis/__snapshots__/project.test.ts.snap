// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`diffFiles works from scratch 1`] = `
{
  ".dockerignore": {
    "data": "# managed by skuba
.gantry/
.git/
.idea/
.serverless/
.vscode/
node_modules*/

/coverage*/
/dist*/
/lib*/
/tmp*/

.DS_Store
npm-debug.log
yarn-error.log
# end managed by skuba
",
    "operation": "A",
  },
  ".eslintignore": {
    "data": "# managed by skuba
.idea/*
.vscode/*

.cdk.staging/
.pnpm-store/
.serverless/
cdk.out/
node_modules*/

/coverage*/
/dist*/
/lib*/
/tmp*/
# end managed by skuba
",
    "operation": "A",
  },
  ".eslintrc.js": {
    "data": "module.exports = {
  extends: ['skuba'],
};
",
    "operation": "A",
  },
  ".github/renovate.json5": {
    "data": "{
  extends: ['github>seek-oss/rynovate'],
}
",
    "operation": "A",
  },
  ".gitignore": {
    "data": "# managed by skuba
.idea/*
.vscode/*
!.vscode/extensions.json

.cdk.staging/
.serverless/
cdk.out/
node_modules*/

/coverage*/
/dist*/
/lib*/
/tmp*/

.DS_Store
.eslintcache
.pnpm-debug.log
*.tgz
*.tsbuildinfo
npm-debug.log
package-lock.json
yarn-error.log
# end managed by skuba
",
    "operation": "A",
  },
  ".prettierignore": {
    "data": "# managed by skuba
# Gantry resource files support non-standard template syntax
/.gantry/**/*.yaml
/.gantry/**/*.yml
gantry*.yaml
gantry*.yml
pnpm-lock.yaml
coverage
# end managed by skuba
",
    "operation": "A",
  },
  ".prettierrc.js": {
    "data": "module.exports = require('skuba/config/prettier');
",
    "operation": "A",
  },
  "jest.config.ts": {
    "data": "import { Jest } from 'skuba';

export default Jest.mergePreset({
  coveragePathIgnorePatterns: ['src/testing'],
  coverageThreshold: {
    global: {
      branches: 0,
      functions: 0,
      lines: 0,
      statements: 0,
    },
  },
  setupFiles: ['<rootDir>/jest.setup.ts'],
  testPathIgnorePatterns: ['/test\\\\.ts'],
});
",
    "operation": "A",
  },
  "jest.setup.ts": {
    "data": "process.env.ENVIRONMENT = 'test';

export {};
",
    "operation": "A",
  },
  "package.json": {
    "data": "{
  "private": true,
  "license": "UNLICENSED",
  "scripts": {
    "build": "skuba build",
    "format": "skuba format",
    "lint": "skuba lint",
    "start": "skuba start",
    "test": "skuba test --coverage",
    "test:watch": "skuba test --watch"
  },
  "skuba": {
    "entryPoint": "src/app.ts",
    "template": null,
    "type": "application",
    "version": "0.0.0-semantically-released"
  }
}
",
    "operation": "A",
  },
}
`;
